RFCs/0027-Merge-Dotnet-Test-And-VsTest-Verbs.md

```diff
@@ -0,0 +1,46 @@
+# 0027 Merge dotnet test and dotnet vstest verb
+
+## Motivation
+Customer ask to merge the two dotnet verbs "test" and "vstest".
+Associated github issue: https://github.com/Microsoft/vstest/issues/1453
+
+## Goal
+User should be able to run unit and integration tests with ability to pass in the filename of either a .csproj or a .dll without having to change any other syntax of the dotnet test command.
+
+## Success criteria (as per user ask)
+1. The "dotnet test" command can be used in place of the "dotnet vstest" command.
+2. The "dotnet test" command can accept a list of dlls
+3. Using "dotnet test" and targeting a list of dlls, there is an argument for "--parallel" which will run the tests from the dlls in parallel.
+4. Using "dotnet test" and targeting a dll, the inclusion or exclusion of the "--no-build" argument will have no impact on the behavior of execution.
+
+## Issues with merging the two verbs
+
+### Irrelevant Arguments
+The arguments that are acceptable for the dotnet test for project/solution will not be applicable to the dotnet test when invoked for dlls.
+Here are possible ways to handle this:
+
+1. Ignore the arguments that are not valid.
+2. Throw a warning to the user, and continue with the execution.
+3. Throw an error, giving user an actionable message and stop the execution.
+
+Note:
+1. For option (1), ignoring the arguments seems incorrect. If the user gives certain arguments, the expectation is that the arguments get honored. Ignoring the arguments without notify the user would be incorrect. If an argument like "--no-build" is not relevant 
+
+### Same arguments but different acceptable values
+There are arguments that are accepted by both dotnet test and dotnet vstest today. We will have make sure there is parity with respect to these arguments before merge.
+One such argument is "--framework". The values accepted by the dotnet test today are different from the ones accepted by dotnet vstest. Test platform today does not understand aliases. For example, ".NETCoreApp,Version=v1.0" is valid framework while netcoreapp1.0 is not valid framework for test platform.
+
+We will have to make sure all such arguments accept the super set of these values now. We might have to add another layer to convert the values into appropriate values that the platform understands.
```

Comment:
".NETCoreApp,Version=1.0" is the concatenation of `$(TargetFrameworkIdentifier),Version=$(TargetFrameworkVersion)`. The short form is `TargetFrameworkMoniker` (TFM), ie "netcoreapp1.0".

Why does dotnet vstest use the long form instead of the TFM as there is a 1:1 mapping between both? If there's no strong reason for that besides historical ones I would recommend to just use the TFM.

Reply:
I guess we should be using https://www.nuget.org/packages/NuGet.Frameworks/ then?
But for that, I guess we first need to move to netstandard2.0

Reply:
@ericstj thoughts?

Reply:
Be careful in terminology here.  `$(TargetFrameworkMoniker)` is the full form and the one understood by the [FrameworkName](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.versioning.frameworkname?view=netframework-4.8) class.  For this there is a 1:1 mapping where a given framework can only be represented by one string.

`$(TargetFramework)` is the NuGet version of this, which doesn't have those characteristics.  Many strings can be used to represent the same framework.  So things like net45 and net4.5 are treated the same.

Reply:
dotnet verbs today take -f or --framework as an argument, and this arguments supposedly is the NuGet version, as it accepts net451 as a valid value.

We on the other hand relied on FrameworkName, what I am thinking of doing is accept the input as NuGet version and then convert that into FrameworkName.
I thought NuGet.Framework has the mapping that I could leverage.

What do you guys suggest ?
The code sits here https://github.com/microsoft/vstest/blob/master/src/Microsoft.TestPlatform.ObjectModel/Framework.cs

Reply:
I guess someone from the CLI team should be able to help here. 

cc @wli3 @livarcocc @nguerrera

Reply:
@livarcocc @nguerrera Is that the right approach ?

Reply:
cc @KathleenDollard 

I cannot recall any CLI command that is not default to short version. I don't think it will accept the longer version in most of the cases, since we usually just forward it to MSBuild as property TargetFramework